src/metadata.rs

```diff
@@ -128,13 +194,237 @@ pub trait Metadata: MajorObject {
         Ok(())
     }
 
+    /// For Datasets this sets the dataset name; normally
+    /// application code should not set the "description" for
+    /// GDALDatasets. For RasterBands it is actually a description
+    /// (if supported) or `""`.
     fn set_description(&mut self, description: &str) -> Result<()> {
-        // For Datasets this sets the dataset name; normally
-        // application code should not set the "description" for
-        // GDALDatasets. For RasterBands it is actually a description
-        // (if supported) or "".
         let c_description = CString::new(description.to_owned())?;
         unsafe { gdal_sys::GDALSetDescription(self.gdal_object_ptr(), c_description.as_ptr()) };
         Ok(())
     }
+
+    /// Get an iterator over metadata entries, across all domains.
+    ///
+    /// # Example
+    ///
+    /// ```rust, no_run
+    /// use gdal::{Dataset, Metadata, MetadataEntry};
+    /// # fn main() -> gdal::errors::Result<()> {
+    /// let dataset = Dataset::open("fixtures/tinymarble.tif")?;
+    /// for MetadataEntry { domain, key, value } in dataset.metadata_iter() {
+    ///   let domain = if domain == "" { "DEFAULT".to_string() } else { domain };
+    ///   println!("{domain}: {key}={value}");
+    /// }
+    /// # Ok(())
+    /// # }
+    /// ```
+    /// Output:
+    /// ```text
+    /// IMAGE_STRUCTURE: INTERLEAVE=PIXEL
+    /// DEFAULT: AREA_OR_POINT=Area
+    /// ...
+    /// DERIVED_SUBDATASETS: DERIVED_SUBDATASET_1_NAME=DERIVED_SUBDATASET:LOGAMPLITUDE:fixtures/tinymarble.tif
+    /// DERIVED_SUBDATASETS: DERIVED_SUBDATASET_1_DESC=log10 of amplitude of input bands from fixtures/tinymarble.tif
+    /// ```
+    fn metadata_iter(&self) -> MetadataIter
```

Comment:
Would it be more idiomatic to call this `metadata`?

Reply:
> 



> Would it be more idiomatic to call this `metadata`?

@lnicola How about just `iter()`?

Reply:
> How about just `iter()`?

Nevermind... since this is a trait, it could/would clash with other `iter`s on the other types.

Reply:
@lnicola I'm a bit stuck not boxing this and yet keeping the return type tidy. 

Of course I really want this:

    fn metadata(&self) -> impl Iterator<MetadataEntry>


But can't until [_Impl trait initiative_](https://rust-lang.github.io/impl-trait-initiative/explainer/rpit_trait.html) happens.

Expanding (and leaking) the _actual_ type leads to this:

    fn metadata(&self) -> FromFn<fn() -> FlatMap<IntoIter<String>, FilterMap<IntoIter<String>, fn(String) -> Option<MetadataEntry>>, fn(String) -> FilterMap<IntoIter<String>, fn(String) -> Option<MetadataEntry>>>>

🤮 

I then tried this:

    fn metadata(&self) -> FromFn<fn() -> Option<MetadataEntry>>

But get the error 

    closures can only be coerced to `fn` types if they do not capture any variables

(and the closure has to capture _something_ to maintain state).

So the best I'm able to come up with is this:

    fn metadata(&self) -> FromFn<Box::<dyn FnMut() -> Option<MetadataEntry> + '_>>

What else should I be considering?

Reply:
There's also this, which I actually like the best because it leaks fewer internals:

    fn metadata(&self) -> Box<dyn Iterator<Item = MetadataEntry> + '_>

Reply:
Oh right, it's a trait.

Reply:
Oof, sorry, but now I'm thinking `impl Iterator` isn't ideal here. Most people seem to avoid it in public APIs, and it probably makes thing like storing the iterator in a struct or implementing more traits (e.g. `DoubleEndedIterator`) harder.

Reply:
> My assumption is because it requires nightly to have a trait with return position impl?

Yes, but `std` doesn't use it outside traits either.

Reply:
So which approach should we choose?

Reply:
I'm inclined to say that the explicit (`MetadataIter`) version is probably better.